File: scala-source/com/barrybecker4/puzzle/slidingpuzzle/ui/SlidingPuzzle.scala
// Copyright by Barry G. Becker, 2017. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.puzzle.slidingpuzzle.ui

import com.barrybecker4.search.Refreshable
import com.barrybecker4.puzzle.common.AlgorithmEnum
import com.barrybecker4.puzzle.common.PuzzleController
import com.barrybecker4.puzzle.common.ui._
import com.barrybecker4.puzzle.common.ui.NavigationPanel
import com.barrybecker4.puzzle.slidingpuzzle.Algorithm
import com.barrybecker4.puzzle.slidingpuzzle.SlidingPuzzleController
import com.barrybecker4.puzzle.slidingpuzzle.model.SlideMove
import com.barrybecker4.puzzle.slidingpuzzle.model.SliderBoard
import com.barrybecker4.ui.util.GUIUtil
import javax.swing._

/**
  * Sliding Puzzle - http://en.wikipedia.org/wiki/Sliding_puzzle.
  * This program solves a difficult classic solitaire puzzle where you slide tiles to make the correct sequence.
  */
object SlidingPuzzle extends App {
  /** Use this to run as an application instead of an applet.  */
  val applet = new SlidingPuzzle(args)
  // this will call applet.init() and start() methods instead of the browser
  GUIUtil.showApplet(applet)
}

case class SlidingPuzzle(myargs: Array[String])
  extends PuzzleApplet[SliderBoard, SlideMove](myargs) with DoneListener {

  private var navPanel: NavigationPanel = _

  /** Construct the application */
  def this() { this(Array[String]()) }

  protected def createViewer = new SliderViewer(this)

  protected def createController(
      viewer: Refreshable[SliderBoard, SlideMove]): PuzzleController[SliderBoard, SlideMove] = {
    new SlidingPuzzleController(viewer)
  }

  protected def getAlgorithmValues: Array[AlgorithmEnum[SliderBoard, SlideMove]] = Algorithm.VALUES

  override protected def createTopControls = SliderTopControls(controller, getAlgorithmValues)

  override protected def createBottomControls: JPanel = {
    navPanel = new NavigationPanel()
    navPanel
  }

  def done(): Unit = {
    navPanel.setPathNavigator(viewer.asInstanceOf[PathNavigator])
  }
}

File: source/com/barrybecker4/puzzle/slidingpuzzle1/ui/SlidingPuzzle.java
// Copyright by Barry G. Becker, 2013. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.puzzle.slidingpuzzle1.ui;

import com.barrybecker4.common.search.Refreshable;
import com.barrybecker4.puzzle.common.AlgorithmEnum;
import com.barrybecker4.puzzle.common.PuzzleController;
import com.barrybecker4.puzzle.common.ui.*;
import com.barrybecker4.puzzle.slidingpuzzle1.Algorithm;
import com.barrybecker4.puzzle.slidingpuzzle1.SlidingPuzzleController;
import com.barrybecker4.puzzle.slidingpuzzle1.model.SlideMove;
import com.barrybecker4.puzzle.slidingpuzzle1.model.SliderBoard;
import com.barrybecker4.ui.util.GUIUtil;

import javax.swing.*;

/**
 * Sliding Puzzle - http://en.wikipedia.org/wiki/Sliding_puzzle.
 * This program solves a very difficult classic solitaire puzzle
 * where you slide tiles to form an image or correct sequence.
 */
public final class SlidingPuzzle extends PuzzleApplet<SliderBoard, SlideMove>
                             implements DoneListener {

    private NavigationPanel navPanel;

    /** Construct the applet */
    public SlidingPuzzle() {}

    /** Construct the application */
    SlidingPuzzle(String[] args) {
        super(args);
    }


    @Override
    protected PuzzleViewer<SliderBoard, SlideMove> createViewer() {
        return new SliderViewer(this);
    }

    @Override
    protected PuzzleController<SliderBoard, SlideMove> createController(Refreshable<SliderBoard, SlideMove> viewer_) {
        return new SlidingPuzzleController(viewer_);
    }

    @Override
    protected AlgorithmEnum<SliderBoard, SlideMove>[] getAlgorithmValues() {
        return Algorithm.values();
    }

    protected SliderTopControls createTopControls() {
        return new SliderTopControls(controller_, getAlgorithmValues());
    }

    @Override
    protected JPanel createBottomControls() {
        navPanel = new NavigationPanel();
        return navPanel;
    }

    @Override
    public void done() {
        navPanel.setPathNavigator((PathNavigator) viewer_);
    }

    /**
     * Use this to run as an application instead of an applet.
     */
    public static void main(String[] args) {

        PuzzleApplet applet = new SlidingPuzzle(args);

        // this will call applet.init() and start() methods instead of the browser
        GUIUtil.showApplet(applet);
    }

}

